Added support for Roidmi Eve

martin9000andersen opened the pull request:
I have added support for robot Roidmi Eve (Robot vacuum cleaner).

Many status values are implemented and many features are implemented e.g.:
    start-clean, stop, go-home, set_sound_volume, set_cleaning_mode, set_sweep_type, set_work_station_freq, start_dust, etc.

I do not know the syntax to activate cleaning of specific rooms, any input on how to make the function "start_room_sweep_unknown" work is highly appreciated.

rytilahti commented on miio/device.py:
```diff
@@ -71,30 +71,22 @@ def accesspoint(self):
-        if self.data["model"] is not None:
-            return self.data["model"]
-        return None
+        return self.data.get("model")
```
Please rebase on top of the current master, #1083 should fix the device info problems.

martin9000andersen replied:
Yes, that commit fixed the problem

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+"""Vacuum Eve Plus (roidmi.vacuum.v60)"""
+
+# https://github.com/rytilahti/python-miio/issues/543#issuecomment-755767331
```
```suggestion
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+    FindChargerPause = 11
+
+
+class RoidmiVacuumStatus(DeviceStatusContainer):
```
```suggestion
class RoidmiVacuumStatus(DeviceStatus):
```
please extend this directly from the DeviceStatus for now. Also, please use the names and return types consistent to vacuumcontainers.py (e.g., `ConsumableStatus` for consumables), as we want to move slowly towards common interfaces among device types for interoperability.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+        return self.data["battery_level"]
+
+    @property
+    def filter_left_time(self) -> int:
```
See my comment above, filter times should use timedeltas.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+        return RoidmiVacuumStatus(
+            {
+                prop["did"]: prop["value"] if prop["code"] == 0 else None
+                # max_properties limmit to 10 to avoid "Checksum error" messages from the device.
```
This might be fixed by the recent #1075 - could you please confirm if that's the case?

martin9000andersen replied:
For Roidmi Eve there are two reasons for getting "checksum error": 1) is too many properties at a time (here I use max_properties), 2) Too many properties in total.
Commit #1075 did not affect No. 1, so I need to keep max_properties, but it did solved No. 2 (This means that I can uncomment all properties in MiotMapping). Thank you very much for telling me about this fix.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+    @property
+    def switch_status(self) -> SwitchStatus:
+        try:
+            return SwitchStatus(self.data["switch_status"])
```
This doesn't currently seem to be queried at all. What does this indicate?

martin9000andersen replied:
I am not sure. It is an Enum and can only have the value "Open". I will remove it from the public interface, and only leave an comment about this in MiotMapping.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+            return SwitchStatus.Unknown
+
+    @property
+    def device_status(self) -> DeviceStatus:
```
```suggestion
    def status(self) -> DeviceStatus:
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+    Idle = 0
+
+
+class FaultStatus(Enum):
```
It is probably simpler to have this as a dictionary than as an enum (the value being a user-friendly message, see https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L16).

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+    def auto_boost(self) -> int:
+        return self.data["auto_boost"]
+
+    def parseForbidMode(self, val):
```
Internal methods should be marked as such. Also, these could probably be moved to be in the scope of `forbid_mode`?

```suggestion
    def _parse_forbid_mode(self, val):
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+        return self.data["auto_boost"]
+
+    def parseForbidMode(self, val):
+        def secToClock(val):
```
```suggestion
        def _secToClock(val):
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+        active = bool(asDict["time"][2])
+        begin = secToClock(asDict["time"][0])
+        end = secToClock(asDict["time"][1])
+        return json.dumps(
+            {"enabled": active, "begin": begin, "end": end, "tz": asDict["tz"]}
+        )
```
Instead of using json dumps, would it be possible to use the `DNDStatus` from `vacuumcontainers.py` for this? 

If not, please create a separate container class (derived from `DeviceStatus`).

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+            _LOGGER.error("Unknown WaterLevel (%s)", self.data["water_level"])
+            return WaterLevel.Unknown
+
+    # @property
+    # def siid8_13(self) -> int:
+    #     return self.data["siid8_13"]
+
+    # @property
+    # def siid8_14(self) -> int:
+    #     return self.data["siid8_14"]
```
```suggestion
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+    def clean_counts(self) -> int:
+        return self.data["clean_counts"]
+
+    # @property
+    # def siid8_19(self) -> int:
+    #     return self.data["siid8_19"]
```
```suggestion
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,590 @@
+        click.argument("end", type=str),
+        click.argument("active", type=bool, required=False, default=True),
+    )
+    def set_forbid_mode(self, begin: str, end: str, active: bool = True):
```
```suggestion
    def set_dnd(self, start_hr: int, start_min: int, end_hr: int, end_min: int, active: bool = True)
```
This would make it similar to vacuum.py's `set_dnd` with the added `active` flag, for future proofing :-)

rytilahti reviewed (changes requested):
Thanks for the PR! I did a first round of reviews and left some specific comments inline, but on generla level:
* Please add descriptive docstrings to all properties and methods belonging to the public API
* Remove commented out code (except for the variable mappings, in case someone wants to look into those later)
* Please add (at least) `is_on`, `is_paused`, `got_error` from https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L125 - the long term goal is to make all vacuum implementations to have a common interface, so this will be useful for that.

martin9000andersen commented:
Thank you very much for your comments. I really like your long term goal of a common interface.
I will have a look at it in the coming days.

martin9000andersen commented:
I have updated the code to be more like vacuumcontainers.py (before it was closer to dreamevacuum_miot.py).

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+        return self.data["mop_present"]
+
+    @property
+    def work_station_freq(self) -> int:
```
```suggestion
    def work_station_frequency(self) -> int:
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+        return self.data["led_switch"]
+
+    @property
+    def lidar_collision(self) -> bool:
```
```suggestion
    def lidar_enabled(self) -> bool:
```
Maybe?

martin9000andersen replied:
Text from the Mi App: "Priority use of radar - When turned on, the robot will use radar as the main detection sensor to help reduce collisions".
When this option is OFF, lidar is still used for mapping and navigation.
I do not think that "lidar_collision" or "lidar_enabler" are a great names. Do you have a better idea now that you know more about this option.

rytilahti replied:
I don't have a better recommendation for this at the moment, so let's keep it as it is.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+        return self.set_property("work_station_freq", work_station_freq)
+
+    @command(click.argument("timing", type=str))
+    def set_timing_unknown(self, timing: str):
```
```suggestion
    def set_timing(self, timing: str):
```
I think there is no need to describe this as unknown?

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+        return self.set_property("timing", timing)
+
+    @command(click.argument("auto_boost", type=bool))
+    def set_auto_boost(self, auto_boost: bool):
```
```suggestion
    def set_carpet_mode(self, auto_boost: bool):
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+
+    @command(click.argument("voice", type=str))
+    def set_voice_unknown(self, voice: str) -> None:
+        """Set voice.
+
+        FIXME: the syntax of voice is unknown
+        """
```
Could this be the name of the setting, e.g., "girl_en"?

martin9000andersen replied:
Yes, I think that "girl_en" can be part of the string. But I also think that it must be json format, and I do not know the json-key. (I have tried {"voice": "girl_en"} without lock)

rytilahti replied:
Okay, let's keep it this way, maybe someone figures it out at some point :-)

rytilahti commented on miio/tests/test_roidmivacuum_miot.py:
```diff
@@ -0,0 +1,188 @@
+        status = self.device.status()
+        assert status.auto_boost == _INITIAL_STATE["auto_boost"]
+        assert status.battery == _INITIAL_STATE["battery_level"]
+        assertEnum(
+            status.charging_state, ChargingState(_INITIAL_STATE["charging_state"])
+        )
```
```suggestion
        assert status.charging_state == ChargingState(_INITIAL_STATE["charging_state"])
```
should be enough for all enum checks. If they are equal, their `repr()`s are also equal.

rytilahti commented on miio/tests/test_roidmivacuum_miot.py:
```diff
@@ -0,0 +1,188 @@
+    request.cls.device = DummyRoidmiVacuumMiot()
+
+
+def assertEnum(a, b):
+    assert a == b
+    assert repr(a) == repr(b)
```
```suggestion
```
This is unnecessary.

rytilahti commented on miio/tests/test_roidmivacuum_miot.py:
```diff
@@ -0,0 +1,188 @@
+        assert status.total_area == _INITIAL_STATE["total_clean_areas"]
+        assert status.count == _INITIAL_STATE["clean_counts"]
+
+    def test_RoidmiConsumableStatus(self):
```
```suggestion
    def test_consumable_status(self):
```
Avoid `CamelCasing`, it is also unnecessary to repeat "Roidmi" in tests.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+    @command()
+    def reset_filter_life(self) -> None:
+        """Reset filter life."""
+        return self.call_action_by(10, 1)
```
Please convert the actions to use `call_action(str)` and declare the actions in the mapping like: https://github.com/rytilahti/python-miio/pull/867/files#diff-2f2ea69d0a6bf409d623cea17e4744ce7b60739d4c29b79166a8c1979660bfb3R36-R42

The `call_action_by` is rather targeted for `miiocli` usage where there is no known symbol for it available.

rytilahti reviewed (commented):
I added some more comments, I think it's good to go after this round but be aware that the API will change for the next major release to allow consolidating the common functionalities among all supported vacuums :-)

While doing the review I did some googling and started to wonder if this PR is for "Roidmi Eve Plus" or is there also a version without the "plus"? If that's the same one, this might be one option when it's time to upgrade away from my gen1 rockrobo.. What's your opinion about the device?

martin9000andersen commented:
> I added some more comments, I think it's good to go after this round but be aware that the API will change for the next major release to allow consolidating the common functionalities among all supported vacuums :-)

Thank you for the comments. I have added comments on a few points, the rest of the points look very clear and I hope to get time to solve them within a week.

martin9000andersen commented:
> While doing the review I did some googling and started to wonder if this PR is for "Roidmi Eve Plus" or is there also a version without the "plus"? If that's the same one, this might be one option when it's time to upgrade away from my gen1 rockrobo.. What's your opinion about the device?

I have the "PLUS" model. The manual covers both models. The robot is the same, the "PLUS" is about the base. The "PLUS" base is with dust-collection. The manual shows a simple base for the non-PLUS model (but I have not seen the non-PLUS version for sale)
The miio protocol has a command to ask about "station-type", that is an int, for me this return 102, and I guess that the simple base will return a different number.

I am very satisfied with Roidmi Eve Plus. The largest draw back I see is missing Home assistant support, but with this pull request and the comming uniform vaccum interface, I think that will be solved.

martin9000andersen commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,702 @@
+
+    @property
+    def led_switch(self) -> bool:
+        """The LED on the robot will be always on."""
```
This is a setting that you can change. I have changed the comment, and I have added a function to set this parameter

rytilahti replied:
Okay, I'd suggest then using `led(self) -> bool` and `set_led(self, bool)` as in elsewhere (like in `airconditioner_miot.py` and `fan_miot.py` among others) to keep consistent naming!

martin9000andersen commented:
I have solved (most of) your comments, but also added a few extra functions.
Let me know if there are more changes you like me to make before merging.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+            return PathMode.Unknown
+
+    @property
+    def mop_present(self) -> bool:
```
```suggestion
    def is_water_box_attached(self) -> bool:
```
is used by rockrobo vacuums for exactly the same thing, right?

martin9000andersen replied:
On Roidme Eve the water-box is integrated in the dust-bin. The mop-module is approximately 1 cm high and can be attached below the combined dust+water-box.
Strictly speaking the mop is only a piece of fabric, and Roidmi Eve does not have a sensor to detect if that is pressent (and I guess that this is the same for Roborock).
If you would like a common name for Roborock and Roidme Eve, then we might be able to argue that the water-box on Roborock is a mop-module?
For now I think that best name for Roidme Eve it "is_mop_module_attached"?

rytilahti replied:
Ah, so this is really about the mop, so I'd just use `is_mop_attached`.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+        return self.data["mop_present"]
+
+    @property
+    def work_station_frequency(self) -> int:
+        """work_station_freq (2 means base dust colect every second time)."""
```
Maybe?
```suggestion
    def dust_collection_frequency(self) -> int:
        """Frequency for emptying the dust bin.
        
            Example: 2 means base the dust bin is emptied every second cleaning.
        """
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+        return self.data["edge_sweep"]
+
+    @property
+    def vacuum_led(self) -> bool:
```
```suggestion
    def led(self) -> bool:
```
this will make it consistent with other integrations. It is enough to have the `station_` prefix below for the station led to differentiate it from this.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+
+    @property
+    def vacuum_led(self) -> bool:
+        """Return if led/display on vaccum is on."""
```
```suggestion
        """Return True if led/display on vaccum is on."""
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+        return self.data["clean_area"]
+
+    @property
+    def progress(self) -> str:
+        """Always return None?"""
```
Is this `None` or `"None"` (as a string)? I'd suggest either removing this if it is not working as expected, as that would otherwise confuse anyone trying to use it (causing issue reports on github for a broken feature :-).

martin9000andersen replied:
I have now tried this again. The returned value is the same as if the feature is in not implemented. I will remove this function, and just leave a comment in the _MAPPING dict

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+    def start_room_sweep_unknown(self, roomstr: str) -> None:
+        """Start room cleaning.
+
+        FIXME: the syntax of room_sweep is unknown
```
Considering `roomstr` is not used at all, does something happen if this command gets executed?

If something happens, I'd move the FIXME to be `_LOGGER.warning()` to let users know about it (maybe someone finds out what the string should be) and otherwise remove the command decorator (so that it's not shown in the mioocli but left here just for a reference).

martin9000andersen replied:
The problem is that I have not figured out how to start an "action" with an argument.
I will comment out all the functions that are calling an action with an argument.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+    def start_sweep_unknown(self, sweep_mode: SweepMode, clean_info: str) -> None:
+        """Start sweep with mode.
+
+        FIXME: the syntax of start sweep with mode is unknown
+        """
```
Is this fixme still relevant considering the parameters are passed to the action?

martin9000andersen replied:
The problem is that I have not figured out how to start an "action" with an argument.
I will comment out all the functions that are calling an action with an argument.

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+        return self.set_property("station_led", on)
+
+    @command(click.argument("on", type=bool))
+    def set_vacuum_led(self, on: bool):
```
```suggestion
    def set_led(self, on: bool):
```

rytilahti commented on miio/roidmivacuum_miot.py:
```diff
@@ -0,0 +1,769 @@
+
+    @command(click.argument("work_station_freq", type=int))
+    def work_station_frequency(self, work_station_freq: int):
+        """Set work_station_freq (2 means base dust colect every second time)."""
```
See my comment above about the getter for this.

rytilahti reviewed (approved):
All set, thanks @martin9000andersen!

kali876 commented:
I see the subject is a bit frozen :D 
I do not hide that it needs the functionality of vacuuming a specific zone.
If you know how to get mapping for zone vacuuming - I can help :)
I can follow all instructions as long as I get them :)
thx
